### PR TITLE
Add OpenSSL compatibility functions for latest version of Apache httpd

### PR DESCRIPTION
**Pull request by haydenroche5**

### CODE REVIEW

File: src/internal.c
Comment (haydenroche5) [inline src/internal.c:9807]: Where OcspEntry is used to represent an OCSP SingleResponse, I've named the variable single.

File: src/internal.c
Comment (haydenroche5) [inline src/internal.c:9873]: Frees the chain of single responses, if they were marked isDynamic. isDynamic isn't set here, so this is a no-op. The dynamically allocated single and status structs get freed directly below.

File: src/internal.c
Comment (haydenroche5) [inline src/internal.c:11851]: This is a mistake. I'll change it back to ssl->heap.

File: src/ocsp.c
Comment (haydenroche5) [inline src/ocsp.c:287]: Elsewhere, we've given the allocation type DYNAMIC_TYPE_OCSP_REQUEST when allocating an OcspResponse. Changed this to match.

File: src/ocsp.c
Comment (haydenroche5) [inline src/ocsp.c:297]: InitOcspRsponse sets newStatus to zero, so this isn't needed.

File: src/ocsp.c
Comment (haydenroche5) [inline src/ocsp.c:508]: This made sense when OCSP CertId was mapped to OcspRequest. Now it's mapped to OcspEntry, and we just check that the issuer hashes and serial number match.

File: src/ocsp.c
Comment (haydenroche5) [inline src/ocsp.c:601]: In OpenSSL compatibility functions, should we allocate all types with DYNAMIC_TYPE_OPENSSL? What are the guidelines around setting the allocation type?

File: src/ocsp.c
Comment (haydenroche5) [inline src/ocsp.c:955]: The behavior of this function differs depending on if *data is null. See https://www.openssl.org/docs/man1.1.0/man3/i2d_OCSP_CERTID.html.

File: src/ocsp.c
Comment (haydenroche5) [inline src/ocsp.c:969]: WOLFSSL_OCSP_CERTID and WOLFSSL_OCSP_SINGLERESP are represented by the same OcspEntry. Hence, just return the argument.

File: src/ocsp.c
Comment (dgarske) [reply src/ocsp.c:601]: Main requirement is to use the same type on XMALLOC and XFREE. Use best judgement on available types. If merits add a new dynamic type.

File: src/ocsp.c
Comment (dgarske) [reply src/ocsp.c:969]: You'll need a cast here.

File: tests/api.c
Comment (dgarske) [inline tests/api.c:34781]: This printf needs to be moved below variable declarations, otherwise some compilers will complain.

File: tests/api.c
Comment (dgarske) [inline tests/api.c:34813]: This printf needs to be moved below variable declarations, otherwise some compilers will complain.

File: tests/api.c
Comment (dgarske) [inline tests/api.c:34842]: This printf needs to be moved below variable declarations, otherwise some compilers will complain.

File: tests/api.c
Comment (dgarske) [inline tests/api.c:37468]: Can you add const for these test vectors?

**Review by dgarske — CHANGES_REQUESTED** [review]
Excellent work on this! Just a few minor items.

File: src/ocsp.c
Comment (dgarske) [inline src/ocsp.c:960]:
```
./configure CC=g++ --enable-apachehttpd && make
...
src/ocsp.c:960:17: error: assigning to 'unsigned char *' from incompatible type 'void *'
        *data = XMALLOC(id->rawCertIdSize, NULL, DYNAMIC_TYPE_OPENSSL);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by haydenroche5** [comment]
Pulled in some other commits by accident, it seems. I'll remove those from this PR and address your comments @dgarske. Thanks.

File: src/ocsp.c
Comment (haydenroche5) [reply src/ocsp.c:969]: Do I need it if WOLFSSL_OCSP_CERTID and WOLFSSL_OCSP_SINGLERESP are the same type? They're both typedefs for OcspEntry.

File: src/ocsp.c
Comment (haydenroche5) [reply src/ocsp.c:960]: Will look into this.

File: tests/api.c
Comment (haydenroche5) [reply tests/api.c:34719]: I don't think I can because this guy gets assigned to certId.rawCertId, which is not const.

**Review by dgarske — APPROVED** [review]
Code review looks great. I've built and tested without any issues using gcc, clang fsanitize, g++ and scan-build. I'd like another set of eye to review (perhaps @ejohnstown ?) before merging. Thanks @haydenroche5 , great work on this.

**Comment by dgarske** [comment]
@ejohnstown Please consider possible ABI issues with this PR. I didn't see any, but would be good to have another set of eyes confirm.

**Comment by ejohnstown** [comment]
I built this with wolfSSL configured as:

    $ ./configure --enable-apachehttpd --enable-postauth --prefix=$PWD/../wolfssl-osp/apache-httpd/httpd/wolfssl

and I configured Apache as:

    $ ./configure --enable-ssl --with-wolfssl=$PWD/wolfssl --disable-shared --enable-mods-static=all --with-libxml2=/usr/local/Cellar/libxml2/2.9.10_2 --with-included-apr

I gave it the direct path to where homebrew installed libxml2.

It still gave me the ct.h include error. The flag HAVE_WOLFSSL is defined in ap_config_auto.h. That file is only really included in ap_config.h, and that isn't included in md_crypt.c. I added it as other md_*.c files do include it. And it compiled.

Then I got a ton of build errors with md_ocsp.c.

File: tests/api.c
Comment (ejohnstown) [inline tests/api.c:34757]: Please use C style comments, not C++.

File: src/ocsp.c
Comment (ejohnstown) [inline src/ocsp.c:529]: The idiom we tend to follow is to ((XMEMCMP() == 0) && (XMEMCMP() == 0). I know one implementation of memcmp returns the first non-zero difference between the inputs, or zero. Bitwise could indicate no-difference when diffs are present.

**Review by ejohnstown — CHANGES_REQUESTED** [review]

**Review by ejohnstown — APPROVED** [review]

**Comment by ejohnstown** [comment]
Assigning to Todd for final review.

File: wolfssl/wolfcrypt/asn.h
Comment (toddouska) [inline wolfssl/wolfcrypt/asn.h:1305]: No C++ comments, some users are still C only.

File: wolfcrypt/src/asn.c
Comment (toddouska) [inline wolfcrypt/src/asn.c:16742]: What if length is < OCSP_DIGEST_SIZE?  Don't we over read in that case?

File: wolfcrypt/src/asn.c
Comment (toddouska) [inline wolfcrypt/src/asn.c:16748]: Same, what if length is smaller than OCSP_DIGEST_SIZE?

File: wolfcrypt/src/asn.c
Comment (toddouska) [inline wolfcrypt/src/asn.c:17499]: What if the results from memcmp are 0, 1, and -2?  Don't we end with a false 0?

File: src/ocsp.c
Comment (toddouska) [inline src/ocsp.c:857]: Why isn't Free needed anymore?

File: src/ocsp.c
Comment (toddouska) [inline src/ocsp.c:873]: How about this free?

File: wolfcrypt/src/asn.c
Comment (haydenroche5) [reply wolfcrypt/src/asn.c:17499]: Yeah, this is wrong. John pointed out the same error with this in a different instance.

File: src/ocsp.c
Comment (haydenroche5) [reply src/ocsp.c:857]: FreeOcspRequest has the effect of freeing the byte buffers pointed to by req->serial and req->url. We then allocate new memory for these buffers and copy the contents of cid's serial and url into req's newly allocated serial and url. It's not clear to me why this free needed to happen in the first place, if we're just going to allocate the memory again. Can't we just overwrite the relevant fields (serial, url) with cid's fields?

File: src/ocsp.c
Comment (haydenroche5) [reply src/ocsp.c:873]: According to the OpenSSL docs: "OCSP_request_add0_id() adds certificate ID cid to req. It returns the OCSP_ONEREQ structure added so an application can add additional extensions to the request. The id parameter MUST NOT be freed up after the operation." It seems to be that freeing cid here would be a contradiction to this, since the docs say explicitly to not free it.

File: wolfcrypt/src/asn.c
Comment (haydenroche5) [reply wolfcrypt/src/asn.c:16742]: Agreed. Will change OCSP_DIGEST_SIZE to length.

**Review by toddouska — APPROVED** [review]

**Pull request merged**
